crates/wasi-http/src/streams_impl.rs

```diff
@@ -61,18 +73,23 @@ impl crate::streams::Host for WasiHttp {
         this: OutputStream,
         buf: Vec<u8>,
     ) -> wasmtime::Result<Result<u64, StreamError>> {
+        let len = buf.len();
         match self.streams.get(&this) {
-            Some(data) => {
-                let mut new = bytes::BytesMut::with_capacity(data.len() + buf.len());
-                new.put(data.clone());
-                new.put(bytes::Bytes::from(buf.clone()));
-                self.streams.insert(this, new.freeze());
+            Some(st) => {
+                if st.closed {
+                    bail!("stream is dropped!");
+                }
+                let new_len = st.data.len() + len;
+                let mut new = bytes::BytesMut::with_capacity(new_len);
```

Comment:
From https://github.com/bytecodealliance/wasmtime/pull/6228#issuecomment-1516661764, as you said, I think I'd just use `BytesMut` and https://docs.rs/bytes/latest/bytes/struct.BytesMut.html#method.extend_from_slice here

This that would be quite a bit shorter, it could probably also make sense to remove the match altogether and instead rely on https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.and_modify (see example there with `or_insert`)

Reply:
Done. I needed to use `.or_default().extend_from_slice(...)` but I think it achieves the same goal (and thanks for the suggestion of implementing `Default` that made it possible :)